File: prototype/TAT-C/matlab/src/Propagator.m
classdef Propagator
    %J2Propagator. Propagates orbital elements using J2 only
    
    properties (Access = private)
        
        % *Spacecraft. The spacecraft to be propagated.
        Spacecraft
        
        % Double. The J2 term for Earth
        gravParamJ2 = 1.0826269e-003;
        % Double. The gravitational parameter of Earth
        gravParamMu = 3.986004415e+5;
        % Double. The equatorial radius of Earth
        bodyRadius = 6.3781363e+003;
        
        % Double. Julian date of the reference orbital elements
        refEpoch
        % AbsoluteDate. The epoch at which new state is requested.
        propEpoch
        
        % Double. The orbital semi-major axis
        orbitSMA
        % Double. The orbital eccentricity
        orbitECC
        % Double. The orbital inclination
        orbitINC
        % Double. The orbital right ascension of ascending node
        orbitRAAN
        % Double. The orbital argument of periapsis
        orbitAOP
        % Double. The orbital true anomaly
        orbitTA
        % Double. THe orbital mean anomaly
        orbitMA
        
        % Double. The drift in mean motion caused by J2.
        meanMotionRate
        % Double. The drift in argument of periapsis caused by J2.
        argPeriapsisRate
        % Double. The drift in right ascension of ascending node from J2
        rightAscensionNodeRate
        % Double. The orbital semilatus rectum.
        semiLatusRectum
        % Double. The orbital mean motion.
        meanMotion
        
    end
    
    methods (Access = public)
        
        function obj = Propagator(Sat)
            % Class constructor.  Requires a spacecraft.
            obj.Spacecraft = Sat;
            obj = obj.SetOrbitState(Sat.GetOrbitState());
            obj = obj.SetReferenceEpoch(Sat.GetOrbitEpoch);
            obj = ComputeOrbitRates(obj);
        end
        
        function obj = SetPhysicalConstants(obj,gravParamMu,gravParamJ2,bodyRadius)
            %  Set physical constants need to model orbits with J2
            obj.gravParamMu = gravParamMu;
            obj.gravParamJ2 = gravParamJ2;
            obj.bodyRadius = bodyRadius;
        end

        function cartState = Propagate(obj,date)
            % Propgate and return cartesian state given AbsoluteDate
            propDuration = (date.GetJulianDate - obj.refEpoch)*86400;
            orbElem = PropOrbElements(obj,propDuration);
            % Get the pointer to the orbit state
            ptrOrbitState = obj.Spacecraft.GetOrbitState();
            ptrOrbitState.SetKeplerianVectorState(orbElem);
            cartState = ptrOrbitState.GetCartesianState();
        end
        
    end
    
    methods (Access = private)
        
        function obj = SetReferenceEpoch(obj,orbitEpoch)
            % Sets the reference epoch for the initial orbital elements
            obj.refEpoch = orbitEpoch.GetJulianDate();
        end
        
        function obj = SetOrbitEpoch(obj,orbitEpoch)
            %  Set orbit epoch given AbsoluteDate object.
            obj.propEpoch = orbitEpoch.GetJulianDate();
        end
        
        function obj = SetOrbitState(obj,orbitState)
            %  Set orbit state given OrbitState object.
            
            %Done at intialization for performance reasons.
            kepElements = orbitState.GetKeplerianState();
            obj.orbitSMA = kepElements(1);
            obj.orbitECC = kepElements(2);
            obj.orbitINC = kepElements(3);
            obj.orbitRAAN = kepElements(4);
            obj.orbitAOP = kepElements(5);
            obj.orbitTA = kepElements(6);
            obj.orbitMA = obj.TrueToMeanAnomaly(obj.orbitTA,obj.orbitECC);
        end
        
        %  Propagate and return cartesian state
        function cartState = PropCartState(obj,propDuration)
            % Propgate and return cartesian state
            orbElem = PropOrbElements(obj,propDuration);
            cartState = obj.OrbElemToCartesian(orbElem);
        end
        
        function orbElements = PropOrbElements(obj,propDuration)
            %  Propagate and return orbital elements
            twopi = 2*pi;
            orbElements = zeros(6,1);
            orbElements(1) = obj.orbitSMA;
            orbElements(2) = obj.orbitECC;
            orbElements(3) = obj.orbitINC;
            orbElements(4) = mod(obj.orbitRAAN + obj.rightAscensionNodeRate*propDuration,twopi);
            orbElements(5) = mod(obj.orbitAOP + obj.argPeriapsisRate*propDuration,twopi);
            newMA = mod(obj.orbitMA + obj.meanMotionRate*propDuration ,twopi);
            orbElements(6) = obj.MeanToTrueAnomaly(newMA,obj.orbitECC);
        end
        
        function meanMotion = MeanMotion(obj)
            %  Computes the orbital mean motion
            meanMotion = sqrt(obj.gravParamMu/obj.orbitSMA^3);
        end
        
        function semiParameter = SemiParameter(obj)
            %  Computes the orbital semi parameter
            semiParameter = obj.orbitSMA*(1-obj.orbitECC*obj.orbitECC);
        end
        
        function nu = MeanToTrueAnomaly(~,M,e)
            % Computes true anomlaly given mean anomaly and eccentricity
            
            %  pick initial guess for E
            if ( -pi < M < 0 )
                E = M - e;
            else
                E = M + e;
            end
            
            %  Iterate until tolerance is met
            tol = 1e-8;
            diff = 1;
            
            while ( diff > tol )
                f = M - E + e*sin(E);
                fprime = 1 - e*cos(E);
                E_new = E + f/fprime;
                diff = abs(E_new - E);
                E = E_new;
            end
            
            cos_E = cos(E);
            sin_nu = sqrt(1 - e*e)*sin(E)/(1-e*cos_E);       %Vallado pg. 214,  Eq. 4-10
            cos_nu = (cos_E - e)/(1 - e*cos_E);              %Vallado pg. 214,  Eq. 4-12
            nu = atan2(sin_nu,cos_nu);
            
            if (nu < 0)
                nu = nu + 2*pi;
            end
            
        end
        
        function eccentricAnomaly = TrueToEccentricAnomaly(~,trueAnomaly,orbitEcc)
            % Computes ecentric anomaly given true anomaly and eccentricty
            
            cosnu = cos(trueAnomaly);
            %Vallado pg. 213,  Eq. 4-9
            sin_E = sqrt(1 - orbitEcc*orbitEcc)*sin(trueAnomaly)/...
                (1+orbitEcc*cosnu);
            cos_E = (orbitEcc+cosnu)/(1+orbitEcc*cosnu);
            eccentricAnomaly = atan2(sin_E,cos_E);
            
            %  make sure E > 0;
            while (eccentricAnomaly < 0)
                eccentricAnomaly = eccentricAnomaly + 2*pi;
            end
            
        end
        
        function meanAnomaly = E2M(~,eccentricAnomaly,orbitEcc)
            % Computes the mean anomaly given eccentric anomaly and eccentricity
            meanAnomaly = eccentricAnomaly - orbitEcc* sin(eccentricAnomaly);
        end
        
        function meanAnomaly = TrueToMeanAnomaly(obj,trueAnomaly,orbitECC)
            % Computes the true anomaly given mean anomaly and eccentricity
            eccentricAnomaly = obj.TrueToEccentricAnomaly(trueAnomaly,orbitECC);
            meanAnomaly = obj.E2M(eccentricAnomaly,orbitECC);
        end
                
        function obj = ComputeOrbitRates(obj)
            %  Compute orbit element rates
            obj = ComputeMeanMotionRate(obj);
            obj = ComputeArgPerRate(obj);
            obj = ComputeRAANRate(obj);
        end
        
        function obj = ComputeMeanMotionRate(obj)
            %  Computes the orbital mean motion Rate
            %  Vallado, 3rd. Ed.  Eq9-41
            n = obj.MeanMotion();
            J2 = obj.gravParamJ2;
            e = obj.orbitECC;
            p = obj.SemiParameter();
            obj.meanMotionRate = n - 0.75*n*J2*(obj.bodyRadius/p)^2*sqrt(1-e^2)*(3*sin(obj.orbitINC)^2 - 2);
        end
        
        function obj = ComputeArgPerRate(obj)
            %  Computes the argument of periapsis rate
            %  Vallado, 3rd. Ed.  Eq9-39
            n = obj.MeanMotion();
            obj.argPeriapsisRate = 3*n*obj.bodyRadius^2*obj.gravParamJ2/...
                4/obj.SemiParameter()^2*(4 - 5*sin(obj.orbitINC)^2);
        end
        
        function obj = ComputeRAANRate(obj)
            %  Computes rate right ascension of node rate
            %  Vallado, 3rd. Ed.  Eq9-37
            n = obj.MeanMotion();
            obj.rightAscensionNodeRate = -3*n*obj.bodyRadius^2*obj.gravParamJ2/...
                2/obj.SemiParameter()^2*cos(obj.orbitINC);
        end
        
    end
    
end

File: prototype/TAT-C/cpp/src/Propagator.cpp
//------------------------------------------------------------------------------
//                           Propagator
//------------------------------------------------------------------------------
// GMAT: General Mission Analysis Tool.
//
// Copyright (c) 2002 - 2015 United States Government as represented by the
// Administrator of the National Aeronautics and Space Administration.
// All Other Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at:
// http://www.apache.org/licenses/LICENSE-2.0.
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
// express or implied.   See the License for the specific language
// governing permissions and limitations under the License.
//
// Author: Wendy Shoan, NASA/GSFC
// Created: 2016.05.06
//
/**
 * Implementation of the propagator class
 */
//------------------------------------------------------------------------------

#include <cmath>            // for INFINITY
#include "gmatdefs.hpp"
#include "Propagator.hpp"
#include "RealUtilities.hpp"
#include "GmatConstants.hpp"
#include "Rmatrix33.hpp"
#include "TATCException.hpp"
#include "StateConversionUtil.hpp"
#include "MessageInterface.hpp"

//------------------------------------------------------------------------------
// static data
//------------------------------------------------------------------------------
// None

//------------------------------------------------------------------------------
// public methods
//------------------------------------------------------------------------------

//---------------------------------------------------------------------------
//  Propagator(Spacecraft *sat)
//---------------------------------------------------------------------------
/**
 * Default constructor for Propagator.
 *
 * @param sat The spacecraft object
 * 
 */
//---------------------------------------------------------------------------
Propagator::Propagator(Spacecraft *sat) :
   sc                     (sat),
   J2                     (1.0826269e-003),
   mu                     (3.986004415e+5),
   eqRadius               (6.3781363e+003),
   refJd                  (GmatTimeConstants::JD_OF_J2000),
   SMA                    (0.0),
   ECC                    (0.0),
   INC                    (0.0),
   RAAN                   (0.0),
   AOP                    (0.0),
   TA                     (0.0),
   MA                     (0.0),
   meanMotionRate         (0.0),
   argPeriapsisRate       (0.0),
   rightAscensionNodeRate (0.0),
   semiLatusRectum        (0.0),
   meanMotion             (0.0)
{
   OrbitState *orbSt = sc->GetOrbitState();
   SetOrbitState(orbSt);
   refJd = sc->GetOrbitEpoch()->GetJulianDate();
   ComputeOrbitRates();
}

//---------------------------------------------------------------------------
//  Propagator(const Propagator &copy)
//---------------------------------------------------------------------------
/**
 * Copy constructor for Propagator.
 *
 * @param copy The propagator to copy
 * 
 */
//---------------------------------------------------------------------------
Propagator::Propagator(const Propagator &copy) :
   sc                     (copy.sc),   // correct?
   J2                     (copy.J2),
   mu                     (copy.mu),
   eqRadius               (copy.eqRadius),
   refJd                  (copy.refJd),
   propEpoch              (copy.propEpoch),
   SMA                    (copy.SMA),
   ECC                    (copy.ECC),
   INC                    (copy.INC),
   RAAN                   (copy.RAAN),
   AOP                    (copy.AOP),
   TA                     (copy.TA),
   MA                     (copy.MA),
   meanMotionRate         (copy.meanMotionRate),
   argPeriapsisRate       (copy.argPeriapsisRate),
   rightAscensionNodeRate (copy.rightAscensionNodeRate),
   semiLatusRectum        (copy.semiLatusRectum),
   meanMotion             (copy.meanMotion)
{
}

//---------------------------------------------------------------------------
//  Propagator& operator=(const Propagator &copy)
//---------------------------------------------------------------------------
/**
 * operator= for Propagator.
 *
 * @param copy The propagator to copy
 * 
 */
//---------------------------------------------------------------------------
Propagator& Propagator::operator=(const Propagator &copy)
{
   if (&copy == this)
      return *this;
   
   sc                     = copy.sc; // correct?
   J2                     = copy.J2;
   mu                     = copy.mu;
   eqRadius               = copy.eqRadius;
   refJd                  = copy.refJd;
   propEpoch              = copy.propEpoch;
   SMA                    = copy.SMA;
   ECC                    = copy.ECC;
   INC                    = copy.INC;
   RAAN                   = copy.RAAN;
   AOP                    = copy.AOP;
   TA                     = copy.TA;
   MA                     = copy.MA;
   meanMotionRate         = copy.meanMotionRate;
   argPeriapsisRate       = copy.argPeriapsisRate;
   rightAscensionNodeRate = copy.rightAscensionNodeRate;
   semiLatusRectum        = copy.semiLatusRectum;
   meanMotion             = copy.meanMotion;
   
   return *this;
}

//---------------------------------------------------------------------------
//  ~Propagator()
//---------------------------------------------------------------------------
/**
 * destructor for Propagator.
 *
 */
//---------------------------------------------------------------------------
Propagator::~Propagator()
{
}

//------------------------------------------------------------------------------
// void SetPhysicalConstants(Real bodyMu, Real bodyJ2,
//                           Real bodyRadius)
//------------------------------------------------------------------------------
/**
 * Sets physical constant values for the Propagator.
 *
 * @param bodyMu     gravitational parameter to use
 * @param bodyJ2 J2  term to use
 * @param bodyRadius radius of the body
 * 
 */
//---------------------------------------------------------------------------
void Propagator::SetPhysicalConstants(Real bodyMu, Real bodyJ2,
                                      Real bodyRadius)
{
   mu       = bodyMu;
   J2       = bodyJ2;
   eqRadius = bodyRadius;
}

//------------------------------------------------------------------------------
// Rvector6 Propagate(const AbsoluteDate &toDate)
//------------------------------------------------------------------------------
/**
 * Propagates to the input time.
 *
 * @param toDate     date to which to propagate
 * 
 */
//---------------------------------------------------------------------------
Rvector6 Propagator::Propagate(const AbsoluteDate &toDate)
{
   // Propgate and return cartesian state given AbsoluteDate
   Real     propDuration = (toDate.GetJulianDate() -
                           refJd)*GmatTimeConstants::SECS_PER_DAY;
   Rvector6 orbElem      = PropagateOrbitalElements(propDuration);
   // Get the pointer to the orbit state
   OrbitState *ptrOrbitState = sc->GetOrbitState();
   ptrOrbitState->SetKeplerianVectorState(orbElem);
   Rvector6 cartState = ptrOrbitState->GetCartesianState();
   return cartState;
}

//------------------------------------------------------------------------------
// protected methods
//------------------------------------------------------------------------------

//------------------------------------------------------------------------------
// void SetOrbitState(OrbitState &orbState)
//------------------------------------------------------------------------------
/**
 * Sets the orbit state on the Propagator.
 *
 * @param orbState     orbit state
 * 
 */
//---------------------------------------------------------------------------
void Propagator::SetOrbitState(OrbitState *orbState)
{
   // Set orbit state given OrbitState object.
   
   // Done at intialization for performance reasons.
   Rvector6 kepElements = orbState->GetKeplerianState();
   SMA  = kepElements(0);
   ECC  = kepElements(1);
   INC  = kepElements(2);
   RAAN = kepElements(3);
   AOP  = kepElements(4);
   TA   = kepElements(5);
   MA   = StateConversionUtil::TrueToMeanAnomaly(TA,ECC);
}


//------------------------------------------------------------------------------
// Rvector6 PropagateOrbitalElements(Real propDuration)
//------------------------------------------------------------------------------
/**
 * Propagates the orbital elements for the specificed duration.
 *
 * @param propDuration     duration over which to propagate
 * 
 */
//---------------------------------------------------------------------------
Rvector6 Propagator::PropagateOrbitalElements(Real propDuration)
{
   // Propagate and return orbital elements
   Rvector6 orbElements; // all zeros by default
   orbElements(0) = SMA;
   orbElements(1) = ECC;
   orbElements(2) = INC;
   orbElements(3) = GmatMathUtil::Mod(RAAN +
                                      rightAscensionNodeRate*propDuration,
                                      GmatMathConstants::TWO_PI);
   orbElements(4) = GmatMathUtil::Mod(AOP + argPeriapsisRate*propDuration,
                                      GmatMathConstants::TWO_PI);
   Real newMA = GmatMathUtil::Mod(MA + meanMotionRate*propDuration,
                                  GmatMathConstants::TWO_PI);
   orbElements(5) = StateConversionUtil::MeanToTrueAnomaly(newMA,ECC);
   return orbElements;
}

//------------------------------------------------------------------------------
// Real MeanMotion()
//------------------------------------------------------------------------------
/**
 * Computes the mean motion.
 *
 * @return Mean Motion
 * 
 */
//---------------------------------------------------------------------------
Real Propagator::MeanMotion()
{
  // Computes the orbital mean motion
   Real meanMotion = GmatMathUtil::Sqrt(mu/(SMA*SMA*SMA));
   return meanMotion;
}

//------------------------------------------------------------------------------
// Real SemiParameter()
//------------------------------------------------------------------------------
/**
 * Computes the semi parameter.
 *
 * @return SemiParameter
 * 
 */
//---------------------------------------------------------------------------
Real Propagator::SemiParameter()
{
   // Computes the orbital semi parameter
   Real semiParameter = SMA * (1-ECC*ECC);
   return semiParameter;
}

//------------------------------------------------------------------------------
// void ComputeOrbitRates()
//------------------------------------------------------------------------------
/**
 * Computes the orbit rates.
 *
 */
//---------------------------------------------------------------------------
void Propagator::ComputeOrbitRates()
{
   // Compute orbit element rates
   ComputeMeanMotionRate();
   ComputeArgumentOfPeriapsisRate();
   ComputeRightAscensionNodeRate();
}

//------------------------------------------------------------------------------
// void ComputeMeanMotionRate()
//------------------------------------------------------------------------------
/**
 * Computes the mean motion rate.
 * 
 */
//---------------------------------------------------------------------------
void Propagator::ComputeMeanMotionRate()
{
   // Computes the orbital mean motion Rate
   // Vallado, 3rd. Ed.  Eq9-41
   Real n      = MeanMotion();
   Real tmpJ2  = J2;
   Real e      = ECC;
   Real p      = SemiParameter();
   Real sinInc = GmatMathUtil::Sin(INC);
   meanMotionRate = n - 0.75*n*tmpJ2*(eqRadius/p)*(eqRadius/p) *
                    GmatMathUtil::Sqrt(1.0-e*e)*(3.0*sinInc*sinInc - 2.0);
}

//------------------------------------------------------------------------------
// void ComputeArgumentOfPeriapsisRate()
//------------------------------------------------------------------------------
/**
 * Computes the argument of periapsis rate.
 * 
 */
//---------------------------------------------------------------------------
void Propagator::ComputeArgumentOfPeriapsisRate()
{
   // Computes the argument of periapsis rate
   // Vallado, 3rd. Ed.  Eq9-39
   Real n = MeanMotion();
   Real s = SemiParameter();
   Real sinInc = GmatMathUtil::Sin(INC);
   argPeriapsisRate = 3.0*n*(eqRadius*eqRadius)*J2/
                      4.0/(s*s)*(4.0 - 5.0*(sinInc*sinInc));
}

//------------------------------------------------------------------------------
// void ComputeRightAscensionNodeRate()
//------------------------------------------------------------------------------
/**
 * Computes the right ascension node rate.
 * 
 */
//---------------------------------------------------------------------------
void Propagator::ComputeRightAscensionNodeRate()
{
   // Computes rate right ascension of node rate
   // Vallado, 3rd. Ed.  Eq9-37
   Real n = MeanMotion();
   Real s = SemiParameter();
   rightAscensionNodeRate = -3.0*n*(eqRadius*eqRadius)*J2/
                             2.0/(s*s)*GmatMathUtil::Cos(INC);
}

// These should go away
//Real       MeanToTrueAnomaly(Real meanAnomaly, Real e);
//Real       TrueToEccentricAnomaly(Real trueAnomaly, Real e);
//Real       EccentricTiMeanAnomaly(eccentricAnomaly, Real e);
//Real       TrueToMeanAnomaly(trueAnomaly, Real e);


